File: test/corpus/propagation/fallible_ref_access.exs
defmodule RustQ.Corpus.Propagation.FallibleRefAccess do
  @moduledoc "Propagate fallible references through deref and decode_as."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec use_term(term()) :: R.nif_result(R.unit())
  defrust use_term(_term) do
    :ok
  end

  @spec decode_first(R.vec(term())) :: R.nif_result(R.path(:String))
  defrust decode_first(args) do
    text = decode_as!(args.first().ok_or(badarg()), R.path(:String))
    {:ok, text}
  end

  @spec deref_first(R.vec(term())) :: R.nif_result(R.unit())
  defrust deref_first(args) do
    term = deref(args.first().ok_or(badarg()))
    use_term(term)
  end

  @spec decode_map_field(term()) :: R.nif_result(R.path(:String))
  defrust decode_map_field(term) do
    value = decode_as!(term.map_get(Atoms.value()), R.path(:String))
    {:ok, value}
  end
end

File: test/corpus/propagation/fallible_ref_access.rs
fn use_term<'a>(_term: Term<'a>) -> NifResult<()> {
    Ok(())
}

fn decode_first<'a>(args: Vec<Term<'a>>) -> NifResult<String> {
    let text = args
        .first()
        .ok_or(rustler::Error::BadArg)?
        .decode::<String>()?;
    Ok(text)
}

fn deref_first<'a>(args: Vec<Term<'a>>) -> NifResult<()> {
    let term = *args.first().ok_or(rustler::Error::BadArg)?;
    use_term(term)
}

fn decode_map_field<'a>(term: Term<'a>) -> NifResult<String> {
    let value = term.map_get(atoms::value())?.decode::<String>()?;
    Ok(value)
}
